### ISSUE TITLE
Local hostnames not resolving on client devices

### ISSUE THREAD

**Issue by justmytwospence** [linuxserver/docker-wireguard#105]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]

------------------------------

## Expected Behavior
Hostnames on the server LAN should resolve.

## Current Behavior
Hostnames on the server LAN do not resolve.

## Steps to Reproduce
1. Start wireguard server with the below docker compose stack.
2. Connect via one of the created peers.

## Environment
**OS:** Raspbian Lite
**CPU architecture:** Raspberry Pi 4b
**How docker service was installed:** hub.docker.com

DNS is set to a local Pi-hole via `/etc/dhcpcd.conf`:

```
static domain_name_servers=127.0.0.1
```

## Command used to create docker container

```yaml
version: "2.1"

networks:

  default:
    name: vpn-net

services:

  wireguard:
    container_name: vpn
    image: linuxserver/wireguard:version-v1.0.20210315
    restart: unless-stopped
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - ALLOWEDIPS=0.0.0.0/0
      - INTERNAL_SUBNET=10.13.13.0
      - PEERDNS=auto
      - PEERS=${PEERS}
      - PGID=1000
      - PUID=1000
      - SERVERPORT=51820
      - SERVERURL=${SERVERURL}
      - TZ=America/Denver
    ports:
      - 51820:51820/udp
    volumes:
      - ./config:/config
      - /lib/modules:/lib/modules
```

## Docker logs

```log
Attaching to vpn
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-envfile: executing... 
[cont-init.d] 01-envfile: exited 0.
[cont-init.d] 10-adduser: executing... 

-------------------------------------
          _         ()
         | |  ___   _    __
         | | / __| | |  /  \ 
         | | \__ \ | | | () |
         |_| |___/ |_|  \__/


Brought to you by linuxserver.io
-------------------------------------

To support the app dev(s) visit:
WireGuard: https://www.wireguard.com/donations/

To support LSIO projects visit:
https://www.linuxserver.io/donate/
-------------------------------------
GID/UID
-------------------------------------

User uid:    1000
User gid:    1000
-------------------------------------

[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 30-config: executing... 
Uname info: Linux dd2d6fcac623 5.10.27-v7l+ #1409 SMP Tue Apr 6 18:26:02 BST 2021 armv7l armv7l armv7l GNU/Linux
**** It seems the wireguard module is already active. Skipping kernel header install and module compilation. ****
**** Server mode is selected ****
**** External server address is set to spencerboucher.com ****
**** External server port is set to 51820. Make sure that port is properly forwarded to port 51820 inside this container ****
**** Internal subnet is set to 10.13.13.0 ****
**** AllowedIPs for peers 0.0.0.0/0 ****
**** PEERDNS var is either not set or is set to "auto", setting peer DNS to 10.13.13.1 to use wireguard docker host's DNS. ****
**** No wg0.conf found (maybe an initial install), generating 1 server and sunfish,precision,xps peer/client confs ****
grep: /config/peer*/*.conf: No such file or directory
[cont-init.d] 30-config: exited 0.
[cont-init.d] 99-custom-scripts: executing... 
[custom-init] no custom files found exiting...
[cont-init.d] 99-custom-scripts: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[#] ip link add wg0 type wireguard
[#] wg setconf wg0 /dev/fd/63
[#] ip -4 address add 10.13.13.1 dev wg0
[#] ip link set mtu 1420 up dev wg0
[#] ip -4 route add 10.13.13.4/32 dev wg0
[#] ip -4 route add 10.13.13.3/32 dev wg0
[#] ip -4 route add 10.13.13.2/32 dev wg0
[#] iptables -A FORWARD -i wg0 -j ACCEPT; iptables -A FORWARD -o wg0 -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
.:53
CoreDNS-1.8.3
linux/arm, go1.16, 4293992
```

**Comment by aptalca** [comment]
1270.0.1 would refer to inside the container, not host

**Comment by justmytwospence** [comment]
Can you rephrase or elaborate, please? I don't understand.

**Comment by justmytwospence** [comment]
The PEERDNS documentation makes it sound like the wireguard connection will use the host machine's DNS, which is set to the 127.0.0.1 on the host, referring to the pi-hole. I know I'm missing something.

**Comment by justmytwospence** [comment]
Ah are you saying that the Wireguard server is seeing the 127.0.0.1, pulling that into the container where it improperly references itself? What would be the correct way to do this, then?

**Comment by justmytwospence** [comment]
I have tried changing that to the explicit IP address of my docker/pihole host, and that completely kills all DNS on the connected peer device.

**Comment by bananflugan** [comment]
Same issue here, running a Pi-Hole as well which is configured explicitly on my docker host, not using localhost as address.

**Comment by Oddadin** [comment]
Just to add, same exact issue as this one. Came here looking to see if I was being dumb but appears to be the same. Also running a pihole. Any DNS from within the container resolves to 127.0.0.1 but the container can ping etc the IP associated with the target named device.

**Comment by aptalca** [comment]
Posting `same issue` is not really useful. 
The setting `auto` just means the client peer conf is set to send dns requests to the wireguard container, in which coredns is listening on port 53. Coredns by default is set to forward dns requests to whatever dns server is set in the container's `/etc/resolv.conf`, which by default (in bridge networking) is `127.0.0.11`, docker's dns service, which should be set to use whatever dns the host is using.

Follow this chain on your setup and see where the breakdown is occurring.

In the case of the original post, setting the dns server for the container to `127.0.0.1` makes coredns forward requests to itself, because in that context, `127.0.0.1` refers to inside the container not on host.

**Comment by bananflugan** [comment]
My Wireguard is working now, this is my final config. Hope this helps.

```
  wireguard:
    image: ghcr.io/linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - PUID=1003
      - PGID=1003
      - TZ=Europe/Stockholm
      - SERVERURL=wireguard.DOMAIN.COM #optional
      - SERVERPORT=51820 #optional
      - PEERS=my_iphone, wife_iphone #optional
      - PEERDNS=auto #optional
      - INTERNAL_SUBNET=10.13.13.0 #optional
      - ALLOWEDIPS=192.168.10.0/24, 10.13.13.1 #only using the VPN for traffic to my LAN
    volumes:
      - /data/docker/wireguard/config:/config
      - /lib/modules:/lib/modules
    ports:
      - 51820:51820/udp
    #sysctls:
      #- net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
    dns:
      - 192.168.100.14 #ip of my pi-hole container (this setting is important)
    networks:
      docker_net:
        ipv4_address: 192.168.100.16```

**Comment by justmytwospence** [comment]
Thanks both. I'm starting to think my issue is on the pi-hole side because now I can ping the host `server` but it doesn't respond to dns queries over wireguard eg `dig wikipedia.org @192.168.1.2`.

**Comment by justmytwospence** [comment]
Well, the more I test the more confused I get. When connected over wireguard I can ssh into the server host to tail the pihole logs. Then I hit the same host with dig and the pihole logs show that it is getting the request, forwarding it upstream and getting the reply, but then on the client device the dig request says that it times out. Does this seem like a wireguard thing, a pihole thing, or maybe even something else like a firewall issure?

**Comment by bananflugan** [comment]
> Well, the more I test the more confused I get. When connected over wireguard I can ssh into the server host to tail the pihole logs. Then I hit the same host with `dig` on the client device and the pihole logs show that it is getting the request, forwarding it upstream and getting the reply. On the client device, however, the dig request says that it times out. Does this seem like a wireguard thing, a pihole thing, or maybe even something else like a firewall issure?

Is the wireguard container on the same docker net as the pi-hole container? If not try and put them on the same network and see if that helps. Also try and set PEERDNS=192.168.1.2

**Comment by justmytwospence** [comment]
I have tried both of those things to no effect. Current configurations (note that 192.168.1.2 is the static IP of the host for both containers):

```
version: "3"

networks:

  default:
    name: vpn-net

  pihole-net:
    external:
      name: pihole-net

services:

  wireguard:
    container_name: wireguard-server
    image: linuxserver/wireguard:version-v1.0.20210315
    restart: unless-stopped
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    dns:
      - 192.168.1.2
    environment:
      ALLOWEDIPS: 0.0.0.0/0
      INTERNAL_SUBNET: 10.13.13.0
      PEERDNS: 192.168.1.2
      PEERS: 1
      PGID: ${PGID}
      PUID: ${PUID}
      SERVERPORT: 51820
      SERVERURL: ${SERVERURL}
      TZ: ${TZ}
    networks:
      - default
      - pihole-net
    ports:
      - 51820:51820/udp
    volumes:
      - ./config:/config
      - /lib/modules:/lib/modules
```

```
version: "3"

networks:

  default:
    name: pihole-net

  traefik-net:
    external:
      name: traefik-net

services:

  pihole:
    container_name: pihole
    image: pihole/pihole:v5.8.1
    restart: unless-stopped
    cap_add:
      - NET_ADMIN
    dns:
      - 127.0.0.1
      - 9.9.9.9
      - 149.112.112.112
    environment:
      ADMIN_EMAIL: ${ADMIN_EMAIL}
      DNS_BOGUS_PRIV: "true"
      DNS_FQDN_REQUIRED: "true"
      DNSMASQ_LISTENING: all
      DNSSEC: "true"
      PIHOLE_DNS_: 9.9.9.9;149.112.112.112
      QUERY_LOGGING: "true"
      REV_SERVER: "true"
      REV_SERVER_CIDR: ${REV_SERVER_CIDR}
      REV_SERVER_DOMAIN: local
      REV_SERVER_TARGET: ${REV_SERVER_TARGET}
      ServerIP: ${ServerIP}
      TEMPERATUREUNIT: c
      TZ: ${TZ}
      VIRTUAL_HOST: pihole.${DOMAIN}
      WEBPASSWORD: ${WEBPASSWORD}
      WEBUIBOXEDLAYOUT: traditional
    labels:
      - traefik.enable=true
      - traefik.http.routers.pihole.middlewares=lan-only
      - traefik.http.services.pihole.loadbalancer.server.port=80
    networks:
      - default
      - traefik-net
    ports:
      - 53:53/tcp
      - 53:53/udp
      - 67:67/udp
      - 82:80/tcp
    volumes:
      - ./pihole:/etc/pihole
      - ./dnsmasq.d:/etc/dnsmasq.d
```

**Comment by justmytwospence** [comment]
Got it working finally. The key for me was to be explicit about the pihole network's subnet, then point the dns of the wireguard container at that networks gateway IP address.

**Comment by caheredia** [comment]
@justmytwospence 


> Got it working finally. The key for me was to be explicit about the pihole network's subnet, then point the dns of the wireguard container at that networks gateway IP address.

Can you share your settings, I'm having the same issues. I've pointed the wireguard container DNS to my router, but I'm not sure what you mean by "pihole subnet"

**Comment by ngocphamm** [comment]
I'm interested in your solution too @justmytwospence.
